File: backend/app/schemas/procedure.py
"""程序相关 schema（api-specification §5.2 / data-model §3.3 / Q176/Q182/Q278）。

Phase 3 = 基础 CRUD + 多版本骨架。upgrade/rollback/deprecate/restore/copy 属 Phase 7。
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import BatchDeleteFailure
from app.schemas.node import ChapterTreeNode, ChapterUpsert, StepOut, StepUpsert

LevelOfUse = Literal["reference", "continuous", "information"]
Status = Literal["DRAFT", "PUBLISHED", "ARCHIVED"]


class ProcedureCreate(BaseModel):
    """创建空白程序（POST /procedures）。code 由后端按 folder.prefix + 序号生成。"""

    folder_id: str
    name: str = Field(min_length=1, max_length=200)
    level_of_use: LevelOfUse  # Q182 必填，无默认
    description: str = Field(default="", max_length=10000)
    risk_level: int = Field(default=1, ge=1, le=5)
    quality_level: int = Field(default=1, ge=1, le=5)
    custom_values: dict[str, Any] = Field(default_factory=dict)


class ProcedureUpdate(BaseModel):
    """更新程序元数据（PUT /procedures/{id}，仅 is_current=true 且 DRAFT）。"""

    name: str = Field(min_length=1, max_length=200)
    level_of_use: LevelOfUse
    description: str = Field(default="", max_length=10000)
    risk_level: int = Field(default=1, ge=1, le=5)
    quality_level: int = Field(default=1, ge=1, le=5)
    custom_values: dict[str, Any] = Field(default_factory=dict)
    version_update_notes: str = Field(default="", max_length=10000)


class ProcedureSaveIn(ProcedureUpdate):
    """编辑器整批保存入参（PUT /procedures/{id}，§17.2）。

    继承程序级元字段；附带脏节点 upsert + 显式删除列表。三者均可空——
    仅传元字段时退化为元信息更新（向后兼容）。新节点用临时 id，后端返回 id 映射。
    """

    chapters: list[ChapterUpsert] = Field(default_factory=list)
    steps: list[StepUpsert] = Field(default_factory=list)
    deleted_chapter_ids: list[str] = Field(default_factory=list)
    deleted_step_ids: list[str] = Field(default_factory=list)


class TransitionIn(BaseModel):
    """状态切换入参（POST /procedures/{id}/transition）。"""

    status: Status
    reason: str = Field(default="", max_length=2000)


# --------------------------------------------------------------------------- #
# 版本管理（Phase 7）
# --------------------------------------------------------------------------- #
class ReasonIn(BaseModel):
    """通用 reason 必填入参（deprecate 等）。"""

    reason: str = Field(min_length=1, max_length=2000)


class RollbackIn(BaseModel):
    """回退入参（POST /procedures/{id}/rollback）。"""

    target_version: int = Field(ge=1)
    reason: str = Field(min_length=1, max_length=2000)


class RestoreIn(BaseModel):
    """恢复入参（POST /procedures/{id}/restore）；原文件夹已删时 target_folder_id 必填。"""

    reason: str = Field(min_length=1, max_length=2000)
    target_folder_id: str | None = None


class CopyIn(BaseModel):
    """复制为新程序入参（POST /procedures/{id}/copy）。"""

    target_folder_id: str
    name: str | None = Field(default=None, max_length=200)


class RestorePreviewOut(BaseModel):
    """恢复前预检查（§22.5）。"""

    folder_exists: bool
    deprecated_from_folder_id: str | None
    folder_full_path: str | None
    version_count: int


class DiscardDraftResult(BaseModel):
    """丢弃 DRAFT 特殊路径响应（§22.11）。"""

    deleted_id: str
    new_current_id: str
    new_current_version: int


class VersionListItem(BaseModel):
    """group 版本列表行（§22.2 / GET /procedure-groups/{group_id}/versions）。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    version: int
    status: str
    is_current: bool
    version_update_notes: str
    version_update_notes_preview: str
    created_at: datetime
    archived_at: datetime | None


class VersionListOut(BaseModel):
    """group 版本列表（count_only=true 时 items 为空仅返 count）。"""

    count: int
    items: list[VersionListItem] = Field(default_factory=list)


class ProcedureDeleteIn(BaseModel):
    """删除入参（reason 必填，Q128）。"""

    reason: str = Field(min_length=1, max_length=2000)


class ProcedureBatchDeleteIn(BaseModel):
    """批量软删（原子，≤100 项，Q20/Q325）。"""

    ids: list[str] = Field(min_length=1, max_length=100)
    reason: str = Field(default="", max_length=2000)


class BatchMoveIn(BaseModel):
    """批量移动到叶子文件夹（code 不变，Q22/Q273）。"""

    ids: list[str] = Field(min_length=1, max_length=100)
    target_folder_id: str


class BatchMoveResult(BaseModel):
    """批量移动结果：成功 ID + 错误详情（原子，任一失败全不动）。"""

    moved_ids: list[str]
    failed: list[BatchDeleteFailure]


class ProcedureOut(BaseModel):
    """程序库列表行（每 group 一行 is_current + derived 字段，Q176）。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    procedure_group_id: str
    code: str
    name: str
    version: int
    is_current: bool
    status: str
    folder_id: str
    folder_full_path: str
    level_of_use: str
    risk_level: int
    quality_level: int
    description: str
    revision: int
    version_count_in_group: int
    created_at: datetime
    updated_at: datetime


class ProcedureMeta(BaseModel):
    """程序详情的 procedure 子对象（GET /procedures/{id}.procedure）。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    procedure_group_id: str
    code: str
    name: str
    version: int
    is_current: bool
    status: str
    folder_id: str
    folder_full_path: str
    description: str
    risk_level: int
    quality_level: int
    level_of_use: str
    custom_values: dict[str, Any]
    version_update_notes: str
    revision: int
    is_read: bool
    read_at: datetime | None
    deprecated_from_folder_id: str | None
    deprecated_at: datetime | None
    archived_at: datetime | None
    version_change_log: list[dict[str, Any]]
    created_at: datetime
    updated_at: datetime


class ProcedureSaveResult(ProcedureMeta):
    """编辑器保存响应：程序元字段（含新 revision）平铺在顶层 + 新建节点的临时→真实 id 映射。"""

    id_map: dict[str, str] = Field(default_factory=dict)


class FieldOut(BaseModel):
    """程序详情面板渲染用的 active 自定义字段。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    key: str
    field_type: str
    required: bool
    options: list[dict[str, Any]]
    sort_order: int


class ProcedureDetail(BaseModel):
    """GET /procedures/{id} 一次拉全部（Q153）：元信息 + 嵌套章节树 + 平铺步骤。"""

    procedure: ProcedureMeta
    chapters: list[ChapterTreeNode] = Field(default_factory=list)
    steps: list[StepOut] = Field(default_factory=list)
    attachments: list[Any] = Field(default_factory=list)
    fields: list[FieldOut] = Field(default_factory=list)

File: frontend/src/types/procedure.ts
// 程序类型（与后端 app/schemas/procedure.py 对齐）

import type { FieldOption, FieldType } from '@/types/field'

export type LevelOfUse = 'reference' | 'continuous' | 'information'
export type ProcedureStatus = 'DRAFT' | 'PUBLISHED' | 'ARCHIVED'

export interface ProcedureRow {
  id: string
  procedure_group_id: string
  code: string
  name: string
  version: number
  is_current: boolean
  status: ProcedureStatus
  folder_id: string
  folder_full_path: string
  level_of_use: LevelOfUse
  risk_level: number
  quality_level: number
  description: string
  revision: number
  version_count_in_group: number
  created_at: string
  updated_at: string
}

export interface ProcedureMeta {
  id: string
  procedure_group_id: string
  code: string
  name: string
  version: number
  is_current: boolean
  status: ProcedureStatus
  folder_id: string
  folder_full_path: string
  description: string
  risk_level: number
  quality_level: number
  level_of_use: LevelOfUse
  custom_values: Record<string, unknown>
  version_update_notes: string
  signoff_enabled: boolean
  revision: number
  is_read: boolean
  read_at: string | null
  deprecated_from_folder_id: string | null
  deprecated_at: string | null
  archived_at: string | null
  version_change_log: Array<Record<string, unknown>>
  created_at: string
  updated_at: string
}

export interface ProcedureFieldView {
  id: string
  name: string
  key: string
  field_type: FieldType
  required: boolean
  options: FieldOption[]
  sort_order: number
  show_on_cover: boolean
}

export interface ProcedureDetail {
  procedure: ProcedureMeta
  attachments: unknown[]
  fields: ProcedureFieldView[]
  has_source_docx: boolean
}

export interface ProcedureCreate {
  folder_id: string
  name: string
  level_of_use: LevelOfUse
  description?: string
  risk_level?: number
  quality_level?: number
  custom_values?: Record<string, unknown>
}

export interface ProcedureUpdate {
  name: string
  level_of_use: LevelOfUse
  description?: string
  risk_level?: number
  quality_level?: number
  custom_values?: Record<string, unknown>
  version_update_notes?: string
  signoff_enabled?: boolean
}

export interface TransitionPayload {
  status: ProcedureStatus
  reason?: string
}

export interface BatchMoveResult {
  moved_ids: string[]
  failed: Array<{ id: string; code: string; message: string }>
}

// ---- 版本管理（Phase 7，与后端 app/schemas/procedure.py 对齐） ---- //

// 丢弃 DRAFT 特殊路径响应（§22.11）。普通软删返 204（前端得 null）。
export interface DiscardDraftResult {
  deleted_id: string
  new_current_id: string
  new_current_version: number
}

// 恢复前预检查（§22.5）。
export interface RestorePreview {
  folder_exists: boolean
  deprecated_from_folder_id: string | null
  folder_full_path: string | null
  version_count: number
}

// group 版本列表行（GET /procedure-groups/{group_id}/versions）。
export interface VersionListItem {
  id: string
  version: number
  status: ProcedureStatus
  is_current: boolean
  version_update_notes: string
  version_update_notes_preview: string
  created_at: string
  archived_at: string | null
}

export interface VersionListOut {
  count: number
  items: VersionListItem[]
}

export interface RollbackPayload {
  target_version: number
  reason: string
}

export interface RestorePayload {
  reason: string
  target_folder_id?: string | null
}

export interface CopyPayload {
  target_folder_id: string
  name?: string | null
}
